index.js

```diff
@@ -11,14 +15,104 @@ function licensee (configuration, path, callback) {
   } else if (!validSPDX(configuration.license)) {
     callback(new Error('Invalid license expression'))
   } else {
-    // Read the package tree from `node_modules`.
+    if (configuration.productionOnly) {
+      // In order to ignore devDependencies, we need to read:
+      //
+      // 1. the dependencies-only dependency graph, from
+      //    `npm ls --json --production`
+      //
+      // 2. the structure of `node_modules` and `package.json`
+      //    files within it, with read-package-tree.
+      //
+      // `npm ls` calls read-package-tree internally, but does
+      // lots of npm-specific post-processing to produce the
+      // dependency tree.  Calling read-package-tree twice, at
+      // the same time, is far from efficient.  But it works,
+      // and doing so helps keep this package small.
+      runParallel({
+        dependencies: readDependencyList,
+        packages: readFilesystemTree
+      }, function (error, trees) {
+        if (error) callback(error)
+        else withTrees(trees.packages, trees.dependencies)
+      })
+    } else {
+      // If we are analyzing _all_ installed dependencies,
+      // and don't care whether they're devDependencies
+      // or not, just read `node_modules`.  We don't need
+      // the dependency graph.
+      readFilesystemTree(function (error, packages) {
+        if (error) callback(error)
+        else withTrees(packages, false)
+      })
+    }
+  }
+
+  function withTrees (packages, dependencies) {
+    callback(null, findIssues(
+      configuration, packages, dependencies, []
+    ))
+  }
+
+  function readDependencyList (done) {
+    var child = spawn(
+      'npm', ['ls', '--production', '--json'], {cwd: path}
+    )
+    var outputError
+    var json
+    simpleConcat(child.stdout, function (error, buffer) {
+      if (error) outputError = error
+      else json = buffer
+    })
+    child.once('close', function (code) {
+      if (code !== 0) {
+        done(new Error('npm exited with status ' + code))
+      } else if (outputError) {
+        done(outputError)
+      } else {
+        parseJSON(json, function (error, graph) {
+          if (error) return done(error)
+          if (!graph.hasOwnProperty('dependencies')) {
+            done(new Error('cannot interpret npm ls --json output'))
+          } else {
+            var flattened = {}
+            flattenDependencyTree(graph.dependencies, flattened)
+            done(null, flattened)
+          }
+        })
+      }
+    })
+  }
+
+  function readFilesystemTree (done) {
     readPackageTree(path, function (error, tree) {
-      if (error) callback(error)
-      else callback(null, findIssues(configuration, tree, []))
+      if (error) return done(error)
+      done(null, tree.children)
     })
   }
 }
 
+var KEY_PREFIX = '.'
```

Comment:
lol, i've always used $ for this

Reply:
`object.$y` but no `object..x`. Need `object['.x']`, which looks as ridiculous as it should.